src/main/java/org/folio/settings/server/storage/LocaleStorage.java

```diff
@@ -0,0 +1,179 @@
+package org.folio.settings.server.storage;
+
+import static java.net.HttpURLConnection.HTTP_OK;
+import static org.folio.settings.server.util.StringUtil.isBlank;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.sqlclient.Tuple;
+import java.util.List;
+import org.folio.okapi.common.SemVer;
+import org.folio.okapi.common.XOkapiHeaders;
+import org.folio.settings.server.data.LocaleSettings;
+import org.folio.tlib.TenantInitConf;
+import org.folio.tlib.postgres.TenantPgPool;
+import org.folio.util.PercentCodec;
+
+public class LocaleStorage {
+
+  private final TenantPgPool pool;
+
+  private final String localeTable;
+
+  /**
+   * Database storage for locale settings for a tenant.
+   */
+  public LocaleStorage(Vertx vertx, String tenant) {
+    this.pool = TenantPgPool.pool(vertx, tenant);
+    this.localeTable = pool.getSchema() + ".locale";
+  }
+
+  /**
+   * Create the database table for the tenant.
+   */
+  public Future<Void> init(TenantInitConf tenantInitConf) {
+    return initTable()
+        .compose(x -> migrateData(tenantInitConf));
+  }
+
+  private Future<Void> initTable() {
+    return pool.execute(List.of(
+        """
+        CREATE TABLE IF NOT EXISTS %s
+          (id uuid NOT NULL,
+           locale text NOT NULL,
+           currency text NOT NULL,
+           timezone text NOT NULL,
+           numberingsystem text NOT NULL)
+        """.formatted(localeTable),
+
+        """
+        DO $$
+          BEGIN
+            INSERT INTO %s
+            VALUES ('d8e0939b-eee9-45e0-942f-b8354335724b',
+                    'en-US',
+                    'USD',
+                    'UTC',
+                    'latn')
+            ON CONFLICT DO NOTHING;
+          EXCEPTION WHEN SQLSTATE 'P0001' THEN NULL;
+          END
+        $$
+        """.formatted(localeTable),
+
+        """
+        CREATE OR REPLACE FUNCTION %s.locale_singleton()
+          RETURNS TRIGGER AS $$
+          BEGIN
+            IF (TG_OP = 'DELETE') THEN
+              RAISE EXCEPTION 'Cannot delete locale settings record.';
+            ELSIF (TG_OP = 'INSERT') THEN
+              RAISE EXCEPTION 'Cannot insert second locale settings record.';
+            END IF;
+            RETURN NEW;
+          END;
+          $$ language plpgsql;
+        """.formatted(pool.getSchema()),
+
+        """
+        CREATE OR REPLACE TRIGGER locale_singleton BEFORE DELETE OR INSERT ON %s
+          EXECUTE FUNCTION %s.locale_singleton()
+        """.formatted(localeTable, pool.getSchema())
+        ));
+  }
+
+  private Future<Void> migrateData(TenantInitConf tenantInitConf) {
+    var moduleFrom = tenantInitConf.moduleFrom();
+    if (moduleFrom != null && moduleFrom.compareTo(new SemVer("1.3.0")) >= 0) {
```

Comment:
If devops use "real" upgrade this will work. But if they disable (old version), then moduleFrom when enabling again (so an upgrade would be missed). 

The only safe solution is probably to always attempt or store that migration has happened. I'm not a devops so I'm not sure about this.

Reply:
Agreed.